feat: table component

deboer-tim opened the pull request:
### What does this PR do?

This is a clean, rebased branch of the table component after previous review comments in this PR.

We have impending technical debt where our expanding number of table components (images, volumes, pods, containers, soon adding Kubernetes objects) will cross with our future design ideas (tables should be sortable, columns configurable, environments groupable, etc). We can't copy/paste these features into every page, and deal with differences between them.

Each page passes an array of objects and Columns to the Table component, which is responsible for creating the basic layout and calling back to the Column renderer to render each cell in the table. The Table uses grid layout and provides enough abstraction that we can implement additional features later. (e.g. optional columns, column reordering, or resizing. grouping. etc)

The Table provides sorting & sort indicators, and tests are included. Support for 'object containers/grouping' in order to support Container list will be done separately since this is already a big PR.

### Screenshot/screencast of this PR

Before:

<img width="983" alt="Screenshot 2023-11-16 at 11 04 40 AM" src="https://github.com/containers/podman-desktop/assets/19958075/51800fef-c327-4416-97b2-32756f869cbd">

After:

<img width="915" alt="Screenshot 2023-11-15 at 12 40 46 PM" src="https://github.com/containers/podman-desktop/assets/19958075/6fb59f70-007e-46dd-893e-95301dc59703">

### What issues does this PR fix or reference?

Fixes #4365.

### How to test this PR?

PR tests; separate PRs for Images and Volumes coming shortly.

lstocchi commented on packages/renderer/src/lib/ImageRow.svelte:
```diff
@@ -0,0 +1,101 @@
+}
+</script>
+
+{#if column === 0}
```
I don't like much this check. If we add a new column between 0 and 1, we need to edit all *Row components.
Wouldn't it be better to use `slots`?

deboer-tim replied:
Good point. I thought of using slots earlier, the problem is that slots are fixed and named - i.e. we need to pick how many and what they're called up front. I'm trying other options, so far I think the best is just using names instead of numbers here.

deboer-tim replied:
Pushed another update, columns are named now (via the column header) instead of numbered. This is makes the code a little more clearer and adding/removing columns in the future is straightforward.

benoitf commented:
hi, from what I've seen, and I think @cdrage did some investigations long time ago, we should use grid system and not table

deboer-tim commented:
Pinged Charlie, he's checked all his old branches and forks and can't find any of the code from that investigation.

I'm somewhat hesitant to switch to grid here since that will hold this up and do 'multiple things' in one PR... but adopting this on a page is also the ideal time to switch, so I'll give it a try.

benoitf commented:
https://github.com/containers/podman-desktop/issues/885

deboer-tim commented:
Updated description, added video, and force-pushed the latest rebased on main. A lot has changed since the last update and would appreciate feedback so that I can address issues before this moves out of draft state.

benoitf commented on packages/renderer/src/lib/image/ImagesList.svelte:
```diff
@@ -214,6 +169,50 @@ function computeInterval(): number {
+  new Column('Size', 'right', '1fr'),
+];
+
+const tableHelper = new (class extends TableHelper {
```
As early feedback I would say I'm not finding convenient this kind of declaration from consumer POV

deboer-tim replied:
I don't love this syntax either. I had an interface before, but I was worried that adding capability in the future (like the previous comparator function) meant that all consumers would be required to change even if they don't care about the functionality yet.

We could stop it at the current capability and extend in other ways. Or maybe switch it to be more like Column where you create an instance and customize via functions? I will keep trying different ways.

deboer-tim replied:
Pushed another commit to change TableHelper to be more like Column definition. Much simpler for clients to create, please take a look.

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -185,6 +163,48 @@ async function fetchUsageData() {
+    return 'Volume is used by a container';
+  }
+
+  compare(column: string): ((object1: any, object2: any) => number) | undefined {
```
should we have dedicated separate objects instead of one helper ?

like one comparator, one selectable

deboer-tim replied:
Selection is per object/row so it needs to stay at the 'helper' level, but I've moved comparators to the columns.

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -185,6 +163,48 @@ async function fetchUsageData() {
+        return au - bu;
+      };
+    } else if (column === 'Size') {
+      // TODO: should sort by value, not alphabetically (e.g. 1 Gb > 2 Mb)
+      //return (a, b) => a.humanSize.localeCompare(b.humanSize);
```
I would expect to not have any, any in signature but a typed field so I know what I can get from there

deboer-tim replied:
Fixed in latest commit.

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -185,6 +163,48 @@ async function fetchUsageData() {
+let table: Table;
+
+const columns: Column[] = [
+  new Column('Status', 'center', '70px'),
```
should we not add the sort comparator on the column object so we have dedicated type there ?

benoitf replied:
```ts
const statusColumn = new Column<ExpectedType>('Status', 'center', '70px');
statusColumn.setSortComparator( and here we can sort using ExpectedType)
```

deboer-tim replied:
Fixed in latest commit.

cdrage reviewed (changes requested):
Amazing! I'm able to actually organize it by age / alphabetical whenever I want. so big 👍 

With regards to table helper, I'm assuming we have to make a new helper for each list scenario for Pods / Containers / etc?

I'm good with the concept of using TableHelper to make it generic and then "extending" it for different use cases. As honestly it seems like the best way to make it as "generic" as possible considering how difficult it is with other tables / etc.

I'm assuming this will scale well with the other lists we are implementing too

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -215,76 +235,14 @@ function gotoCreateVolume(): void {
+      objects="{volumes}"
+      columns="{columns}"
+      row="{VolumeRow}"
+      helper="{tableHelper}">
```
I'm not sure about the helper name, it looks too generic

deboer-tim replied:
Not sure what else to call it? 🤷🏻‍♂️ Let me know if you have any suggestions.

benoitf replied:
maybe some dedicated methods instead of a generic one. (one entry per helper method/goal) ?

deboer-tim replied:
As per other comment, TableHelper is gone in latest commit.

cdrage commented on packages/renderer/src/lib/image/ImagesList.svelte:
```diff
@@ -214,6 +169,50 @@ function computeInterval(): number {
+  }
+
+  compare(column: string): ((object1: any, object2: any) => number) | undefined {
+    if (column === 'Status') {
```
switch statement instead of what if may be better / cleaner?

deboer-tim replied:
Used alternate suggestion to move comparator (and renderer) to column level.

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -215,76 +235,14 @@ function gotoCreateVolume(): void {
+    <Table
+      bind:this="{table}"
+      bind:selectedItemsNumber="{selectedItemsNumber}"
+      objects="{volumes}"
```
data instead of objects ?

deboer-tim replied:
Fixed in latest commit.

cdrage commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -185,6 +163,48 @@ async function fetchUsageData() {
+  }
+
+  compare(column: string): ((object1: any, object2: any) => number) | undefined {
+    if (column === 'Status') {
```
switch statement instead?

deboer-tim replied:
Moved the comparator to each Column as suggested in another comment, no more if/switch.

benoitf commented on packages/renderer/src/lib/volume/VolumeRow.svelte:
```diff
@@ -0,0 +1,34 @@
+import VolumeActions from './VolumeActions.svelte';
+import type { VolumeInfoUI } from './VolumeInfoUI';
+
+export let object: any;
```
I don't see why it should be with any type

deboer-tim replied:
Fixed in latest commit.

benoitf commented on packages/renderer/src/lib/volume/VolumesList.svelte:
```diff
@@ -215,76 +235,14 @@ function gotoCreateVolume(): void {
+      bind:selectedItemsNumber="{selectedItemsNumber}"
+      objects="{volumes}"
+      columns="{columns}"
+      row="{VolumeRow}"
```
if name is row, I would expect that I provide a row entry to the table and then my table has only one row
here it looks it's more like a template here that we use against all the data

benoitf replied:
thinking more on the topic, I'm still wondering if we need this template per row where we need to go through columns
if we add a column we should not forget to customize the row template --> it'll lead to errors

as inside the row, you're checking the column information

so, probably we could set the component per column with proper type as well

ageColumn.setRenderingTemplate(AgeTemplate)

(and it's in charge of only rendering the age)

deboer-tim replied:
Moved renderer to each column in latest commit. This leads to a bunch of small components, but it removes a big if/switch statement and IMHO makes the code a little cleaner and more obvious. Thanks for the suggestion.

deboer-tim commented:
Added new commit:

Lots of changes based on review comments:
- Added a Type to Column and moved the comparator to it. This makes the Column[] creation a little longer, but the typing works well and keeps the TableHelper minimal.
- Renamed 'objects' to 'data'.
- Added the type to ImageRow/VolumeRow - but these are about to change drastically in the next step...
- Florent had a comment about trying ageColumn.setRenderingTemplate() similar to setComparator(). I had thought about this before and was worried about lots of tiny components, but gave it a shot. Many of these are indeed tiny, but it does provide nice componentization (possible to share some of these between types later?) and gets rid of the big if/switch block.
- Every column needs a renderer - so removed the function and just made it part of the constructor for simpler code.
- At this point it seemed like actions should just be a regular column too. Making this change simplified things a little.

Overall I like these changes; simpler implementation and easier/safer to use for consumers. I am still having trouble with the grid layout misbehaving though.

benoitf commented on packages/renderer/src/lib/image/ImagesList.svelte:
```diff
@@ -214,6 +173,32 @@ function computeInterval(): number {
+let selectedItemsNumber: number;
+let table: Table;
+
+let statusColumn = new Column<ImageInfoUI>('Status', 'center', '70px', ImageColumnStatus);
```
I think we'll reach soon a limit on constructor parameters

so while the name of the column looks quite important, I think the other parameters are kind of optional

so either we move to dedicated method or we provide JSON like parameters with an interface so it'll easier to grow/add mandatory parameters/optional parameters

```ts
new Column<ImageInfoUI>('Status', {align: 'center', width: '70px'});
```
or 

```ts
new Column<ImageInfoUI>('Status')
column.width = '70px';
column.align = 'center';
```

or use fluent API like
```ts
Column.name('Status').width('70px').align('center').create();
```

deboer-tim replied:
I went with the first option in the latest commit. It does complicate things a bit because it makes sense to have most of the parameters optional, but that makes more sense and keeps the use simpler.

benoitf reviewed (commented):
much more simple, nice work
I still believe we might simplify again as number of parameters will grow and we won't be able to scale easily or without like lots of `param1, undefined, undefined, param2` like in parameters at some points

lstocchi commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -0,0 +1,148 @@
+<div class="w-full" class:hidden="{data.length === 0}" role="table">
+  <!-- Table header -->
+  <div
+    class="grid grid-flow-row grid-cols-[{gridCol}] gap-x-0.5 mx-5 h-7 sticky top-0 bg-charcoal-700 text-xs text-gray-600 font-bold uppercase z-[2]"
```
You cannot do `grid-cols-[{gridCol}]` or the style will not be applied. You always have to use static class https://tailwindcss.com/docs/content-configuration#dynamic-class-names
You have to pass the class already defined or think some other way to do it

before
![image](https://github.com/containers/podman-desktop/assets/49404737/bf95c63d-8b10-44bd-9a12-d6def96c892d)
after
![image](https://github.com/containers/podman-desktop/assets/49404737/310a0fbe-0636-4654-9024-9b846170e585)

deboer-tim replied:
Thank you! I was getting really confused because the behaviour was inconsistent, and had forgotten this should not work. Latest commit has a fix setting the style directly on the element.

lstocchi reviewed (commented):
I tried it and looks fine. Waiting to being ready to play with it more.

benoitf commented on packages/renderer/src/lib/table/TestTable.svelte:
```diff
@@ -13,21 +13,23 @@ const people: object[] = [
 const columns: Column<any>[] = [nameCol, ageCol];
 
-const tableHelper = new TableHelper<any>('people', person => person.age < 50, 'People over 50 cannot be deleted');
+const row = new Row<any>();
```
shouldn't be type to {name: string; age: string} ?

deboer-tim replied:
Added Person type in latest commit.

deboer-tim commented on packages/renderer/src/lib/image/ImagesList.svelte:
```diff
@@ -214,6 +173,32 @@ function computeInterval(): number {
+  new Column<ImageInfoUI>('Actions', 'right', '150px', ImageColumnActions),
+];
+
+const tableHelper = new TableHelper<ImageInfoUI>('image', image => image.inUse, 'Image is used by a container');
```
I went with a Row class instead of TableHelper in the second latest commit, and moved the 'kind' as a standalone property. The Row is only for setting selected/disabled text for now. I debated whether selection should be a completely standalone thing, but I think using Row will lead to simple use long term as we add other options.

deboer-tim commented:
Clean branch (without Images/Volumes) pushed, PR comment updated, and draft status removed. Ready for final review.

feloy commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -0,0 +1,165 @@
+  let columnWidths: string[] = ['20px', '32px'];
+
+  // custom columns
+  for (const column of columns) {
+    if (column.info.width) {
+      columnWidths.push(column.info.width);
+    } else {
+      columnWidths.push('1fr');
+    }
+  }
```
You could use a map here:

(not tested)
```
columnWidths = columns.map(c => c.info.width ?? '1fr');
```

deboer-tim replied:
Yup, updated in latest commit.

feloy commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -0,0 +1,165 @@
+  if (!row.selectable) {
+    return;
+  }
+  const toggleData = data;
+  toggleData.filter(object => row.selectable?.(object)).forEach(object => (object.selected = checked));
+  data = toggleData;
```
What is the purpose of using the `toggleData` temp variable? As toggleData is a reference to data, it does not seem necessary to me

deboer-tim replied:
You can see this was actually unchanged from before. Svelte only detects changed objects, not internal state, so sometimes you have to explicitly assign something to see the change visually. I did some testing and this doesn't seem to be required anymore, so I removed it.

The same thing existed further down on line 64 with sorting. This one is required, but I can see the same question arising so I changed it to an explicit 'data = data.sort()', and added the required no-mutation comment. I don't think this is any better code-wise, but at least it's obvious it was intentional.

feloy replied:
Now that you tell me the reason, I remember I effectively had to write some `data = [...data]` in some other PR, to make the refresh happen. Not sure what is the recommended way to handle this

feloy replied:
In this article, it is explained that `data = data` is enough to trigger the Svelte changes.

https://javascript.plainenglish.io/update-arrays-and-objects-in-svelte-40c6a84f8a28

deboer-tim replied:
I've seen it elsewhere too so it seem to be a common way of handling it. For now though sort is the only place that needs it and eslint comment make it clear the code is intentional.

feloy commented on packages/renderer/src/lib/table/TestTable.svelte:
```diff
@@ -0,0 +1,40 @@
+  { name: 'Charlie', age: 43 },
+];
+
+const nameCol: Column<Person> = new Column('Name', { width: '3fr', renderer: TestColumnName });
```
Is it really Column&lt;Person>, or Column&lt;string>, for the name only?

feloy commented on packages/renderer/src/lib/table/TestTable.svelte:
```diff
@@ -0,0 +1,40 @@
+const nameCol: Column<Person> = new Column('Name', { width: '3fr', renderer: TestColumnName });
+nameCol.setComparator((a, b) => a.name.localeCompare(b.name));
+
+const ageCol: Column<Person> = new Column('Age', { align: 'right', renderer: TestColumnAge });
```
Column&lt;number>?

feloy commented on packages/renderer/src/lib/table/TestTable.svelte:
```diff
@@ -0,0 +1,40 @@
+const ageCol: Column<Person> = new Column('Age', { align: 'right', renderer: TestColumnAge });
+ageCol.setComparator((a, b) => a.age - b.age);
+
+const columns: Column<any>[] = [nameCol, ageCol];
```
```suggestion
const columns: Column<Person>[] = [nameCol, ageCol];
```

deboer-tim replied:
Fixed in commit.

feloy commented on packages/renderer/src/lib/table/table.ts:
```diff
@@ -0,0 +1,83 @@
+   *
+   * @param comparator
+   */
+  setComparator(comparator: (object1: Type, object2: Type) => number) {
+    this.comparator = comparator;
+  }
```
Did you consider passing this `comparator` as part of the `ColumnInformation`?
For the user, it would be more *functional*.

deboer-tim replied:
tbh I thought passing the Type down through ColumnInformation would be an issue, but seems to be fine in TypeScript so I did this in latest commit - everything goes into ColumnInfo and can be done in the constructor.

feloy commented on packages/renderer/src/lib/table/table.ts:
```diff
@@ -0,0 +1,83 @@
+   * @param selectable a function that returns false when the object should not be selectable
+   * @param disabledText text to display as a tooltip when selection is disabled
+   */
+  setSelectable(selectable: (object: Type) => boolean, disabledText: string) {
+    this.selectable = selectable;
+    this.disabledText = disabledText;
+  }
```
same here, `selectable` could be part of some `RowInformation`. `disabledText` still being set during the construction, depending on the `selectable` option.

deboer-tim replied:
Yes, since Type isn't an issue and it makes sense to have everything the same, I've added RowInformation and done this. The one point I don't love is that it means selectable can be set and not disabledText or vice versa, but that's extremely minimal and doesn't really matter.

Since selection is optional I finally went ahead and made the entire column disappear if selection isn't allowed. (previously the blank column would be there but no checkboxes)

lstocchi commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -0,0 +1,161 @@
+// number of selected items in the list
+export let selectedItemsNumber: number = 0;
+$: selectedItemsNumber = row.info.selectable
+  ? data.filter(object => row.info.selectable?.(object)).filter(object => object.selected).length
```
any reason why you don't have ` data.filter(object => row.info.selectable?.(object) && object.selected)`

deboer-tim replied:
None at all, I just copied from ImagesList and VolumesList and didn't think to fix. :) I'll do it in follow-up PR.

lstocchi reviewed (approved):
LGTM

feloy commented on packages/renderer/src/lib/table/table.ts:
```diff
@@ -0,0 +1,82 @@
+ * A table row.
+ */
+export class Row<Type> {
+  constructor(readonly info: RowInformation<Type>) {}
```
You could raise an error in the constructor (intended for the developer) if selectable and disabledText are both not set.

deboer-tim replied:
Actually, I retract my previous point :) b/c I think PodsList is the reason not to do this: all pods are selectable, so forcing PodsList to provide disabledText actually doesn't make sense.

Technically it still doesn't make sense if someone provides disabledText and not selectable (i.e. nothing can be selected but you provide text for that case), but that seems very marginal and not something we need to protect against.
